### PR TITLE
Merges Zone and ZoneInfo. Adds implementation of Dns.

### LINKED ISSUE #605: Merge Zone and ZoneInfo
Similarly to what was done ProjectManager, we will merge the functional and info objects here.

### LINKED ISSUE #619: Rename getProjectInfo() to getProject()
Rename getProjectInfo() to getProject() in Dns service class and the entire stack.

### PR DESCRIPTION
**Pull request by mderka**
Made functional Zone a subclass of ZoneInfo.
Fixes #605.

Also renamed `getProjectInfo()` to `getProject()` as per suggestion of @mziccard.
Fixes #619.

Added default implementation of Dns.

### CODE REVIEW

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsException.java [hunk 1]
```diff
@@ -31,5 +33,21 @@ public DnsException(IOException exception) {
...
+   * always throw an exception.
+   *
+   * @throws DnsException when {@code ex} was caused by a {@code DnsException}
+   * @throws RetryInterruptedException when {@code ex} is a {@code RetryInterruptedException}
+   */
+  static DnsException translateAndThrow(RetryHelperException ex) {
+    BaseServiceException.translateAndPropagateIfPossible(ex);
+    throw new DnsException(UNKNOWN_CODE, ex.getMessage());
+  }
+
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsException.java]: regarding the comment bellow, though doc not in par with other services I think there is something that we can start using (and refer to in the javadoc) -> https://cloud.google.com/dns/troubleshooting

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java [hunk 2]
```diff
@@ -37,7 +37,7 @@
     @Override
     public Dns create(DnsOptions options) {
       // TODO(mderka) Implement when DnsImpl is available. Created issue #595.
-      return null;
+      return new DnsImpl(options);
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java]: can comment be removed now?

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java [hunk 3]
```diff
@@ -49,9 +49,55 @@
...
+     */
+    abstract Builder nameServers(List<String> nameServers);
+
+    /**
+     * Builds the instance of {@code ZoneInfo} based on the information set by this builder.
+     */
+    public abstract ZoneInfo build();
+  }
+
+  public static class BuilderImpl extends Builder {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java]: This class should be package scope.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java [hunk 4]
```diff
@@ -240,7 +266,7 @@ public Builder toBuilder() {
   }
 
   static ZoneInfo fromPb(com.google.api.services.dns.model.ManagedZone pb) {
-    Builder builder = new Builder(pb.getName());
+    Builder builder = new BuilderImpl(pb.getName());
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java]: I just realized that `fromPb` and `toPb` are not symmetrical in respect to nameServers.
I also think that a better API will return an empty list (and not null), however code could be change
to fix fromPb/toPb

Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java]: Also, you need to change the equals method to be symmetric.
See Storage.BucketInfo#equals as example.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java [hunk 5]
```diff
@@ -33,20 +34,87 @@
  * @see <a href="https://cloud.google.com/dns/zones/">Google Cloud DNS managed zone
  * documentation</a>
  */
-public class Zone implements Serializable {
+public class Zone extends ZoneInfo {
 
-  // TODO(mderka) Zone and zoneInfo to be merged. Opened issue #605.
+  private static final long serialVersionUID = 564454483894599281L;
+  private transient Dns dns;
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: You need to keep the options, consider it in equals/hashCode and implement readObject for assign a value to dns based on the options. See [Bucket](https://github.com/GoogleCloudPlatform/gcloud-java/blob/cdec697470a3e55066936d7099a0f066ae5f3005/gcloud-java-storage/src/main/java/com/google/gcloud/storage/Bucket.java#L406) as an example.

Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: Also, I think we should already have the Serialization test which should include Zone (as well as ZoneInfo).

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+      return new PageImpl<>(new ZonePageFetcher(serviceOptions, cursor, optionsMap),
+          cursor, zones);
+    } catch (RetryHelperException e) {
+      throw DnsException.translateAndThrow(e);
+    }
+  }
+
+  private static Page<ChangeRequest> listChangeRequests(final String zoneName,
+      final DnsOptions serviceOptions, final Map<DnsRpc.Option, ?> optionsMap,
+      Function<Change, ChangeRequest> TRANSFORM_FUNCTION) {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: if you think there is a value in passing TRANSFORM_FUNCTION to this method (as oppose to referring to it directly from the method) then change the param to a valid local-variable/param name.

Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: Same goes to `listDnsRecords`.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java (hunk 2 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java]: The comment for sure can go. However, I want to verify that this class is complete before closing the issue itself.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsException.java (hunk 1 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsException.java]: Agreed. I know about it and I am more waiting for a confirmation that the list is complete. I will be working with it when doing #593.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java (hunk 2 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsOptions.java]: Done. The class seems to be complete. #595 can be closed too.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java (hunk 3 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java]: Fixed.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java (hunk 4 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/ZoneInfo.java]: Fixed both.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java (hunk 5 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: Added options. Opened #631 for the serialization test.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 6 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: It makes the code more readable. I fixed the name.

**Comment by mderka** [comment]
Addressed all comments so far. Ready for more attention.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Dns.java [hunk 7]
```diff
@@ -429,7 +429,7 @@ public static ChangeRequestListOption sortOrder(SortingOrder order) {
    * @see <a href="https://cloud.google.com/dns/api/v1/managedZones/create">Cloud DNS Managed Zones:
    * create</a>
    */
-  ZoneInfo create(ZoneInfo zoneInfo, ZoneOption... options);
+  Zone create(ZoneInfo zoneInfo, ZoneOption... options);
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Dns.java]: javadoc needs an update

Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Dns.java]: Fixed.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+
+  private static Page<Zone> listZones(final DnsOptions serviceOptions,
+      final Map<DnsRpc.Option, ?> optionsMap) {
+    // define transformation function
+    // this differs from the other list operations since zone is functional and requires dns service
+    Function<ManagedZone, Zone> pbToZoneFunction = new Function<ManagedZone, Zone>() {
+      @Override
+      public Zone apply(
+          com.google.api.services.dns.model.ManagedZone zonePb) {
+        return new Zone(serviceOptions.service(), ZoneInfo.fromPb(zonePb));
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: `Zone.fromPb(this, zonePb)`?

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java [hunk 9]
```diff
@@ -0,0 +1,335 @@
...
+      }
+    };
+    try {
+      // get a list of managed zones
+      DnsRpc.ListResult<ManagedZone> result =
+          runWithRetries(new Callable<DnsRpc.ListResult<ManagedZone>>() {
+            @Override
+            public DnsRpc.ListResult<ManagedZone> call() {
+              return serviceOptions.rpc().listZones(optionsMap);
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: like other Impl, rpc could be stored as a member variable.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 8 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: Fixed.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+          ? ImmutableList.<DnsRecord>of()
+          : Iterables.transform(result.results(), transformFunction);
+      return new PageImpl<>(new DnsRecordPageFetcher(zoneName, serviceOptions, cursor, optionsMap),
+          cursor, records);
+    } catch (RetryHelperException e) {
+      throw DnsException.translateAndThrow(e);
+    }
+  }
+
+  DnsImpl(DnsOptions options) {
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: conventionally, constructors are defined before methods.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java [hunk 11]
```diff
@@ -0,0 +1,335 @@
...
+    try {
+      com.google.api.services.dns.model.Change answer =
+          runWithRetries(
+              new Callable<com.google.api.services.dns.model.Change>() {
+                @Override
+                public com.google.api.services.dns.model.Change call() {
+                  return dnsRpc.getChangeRequest(zoneName, changeRequestId, optionsMap);
+                }
+              }, options().retryParams(), EXCEPTION_HANDLER);
+      return answer == null ? null : fromPb(answer); // should never be null
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: comments seems wrong here.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java [hunk 12]
```diff
@@ -33,20 +36,90 @@
...
+    return new Builder(this);
+  }
 
   /**
    * Constructs a {@code Zone} object that contains the given {@code zoneInfo}.
    */
   public Zone(Dns dns, ZoneInfo zoneInfo) {
-    this.zoneInfo = checkNotNull(zoneInfo);
-    this.dns = checkNotNull(dns);
+    super(new BuilderImpl(zoneInfo));
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: package scope, though preferably implement `Zone.fromPb(Dns, zonePb)` and use that in `DnsImpl.listZones`

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java [hunk 13]
```diff
@@ -61,8 +134,7 @@ public Zone(Dns dns, ZoneInfo zoneInfo) {
   public static Zone get(Dns dnsService, String zoneName, Dns.ZoneOption... options) {
     checkNotNull(zoneName);
     checkNotNull(dnsService);
-    ZoneInfo zoneInfo = dnsService.getZone(zoneName, options);
-    return zoneInfo == null ? null : new Zone(dnsService, zoneInfo);
+    return dnsService.getZone(zoneName, options);
```
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: I think we decided not to have this method as it does not add any value to `dns.getZone`.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 9 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: As per discussion, kept consistent. However, pulled rpc() call out of the retries using a local variable.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 10 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: Moved up

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 11 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: true, removed

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java (hunk 12 above)
Comment (aozarov) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: Oh, you already have that toPb, so I think we should get rid of this method and update listZones as suggested

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java (hunk 9 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/DnsImpl.java]: Fixed.

File: gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java (hunk 13 above)
Comment (mderka) [inline gcloud-java-dns/src/main/java/com/google/gcloud/dns/Zone.java]: Removed.

**Pull request merged**